### PR TITLE
[knx] Add integration tests

### PR DESCRIPTION
**Pull request by holgerfriedrich**
* Add integration tests which include decoding and encoding KNX frames, and the handling by Calimero library as well. This ensures that the KNX wire format is handled properly.
* Add itests for *all* main DPT numbers supported by Calimero.
* Add a mechanism to check for new missing DPT main numbers when new Calimero release is integrated.

The idea of this PR is basically to check the whole way from bytes on the KNX bus through Calimero, into the plugin, and back again via Calimero down to the bus again. The implementation overwrites KNXNetworkLink to implement a test interface. Calimero is then set up on top.

The benefit of this approach is that we now cover large parts of the handling of KNX frame in the Calimero library, without the need of setting up real itests including bus communication to mimic the KNX communication.

PR is ready for review, tests for more DPT subtypes can be added later.

### CODE REVIEW

**Comment by J-N-K** [comment]
Maybe we can test both at once and return the send frame as incoming frame from the bus?

**Comment by holgerfriedrich** [comment]
> Maybe we can test both at once and return the send frame as incoming frame from the bus?

Exactly. Now a CEMI frame is generated from the data, which is used to send an indication to all listeners. A ```ProcessCommunicator``` impl gets the data and supplies it to a dummy ```ProcessListener```, where the test can grab the information. This is close to the way we supply data from KNX to ```ValueDecoder.decode()```.

Concept is working, test for 251.600 is there.

PR needs some generalization and other data types to be covered.

**Comment by holgerfriedrich** [comment]
Generalization to other DPTs is done, tests for different DPTs have been added.
Extending the tests to more DPTs is not scope of this PR.

**Comment by holgerfriedrich** [comment]
Cleaned up the commit history, as this should not be squashed during merge.
Ready for review.

**Comment by holgerfriedrich** [comment]
Fixed a few typos in the code comments. Ready for review.

**Comment by holgerfriedrich** [comment]
@J-N-K can I ask you to have a short look into this PR?
Maybe you can remember why we had the code handling "no-day" for DPT 19.001. I removed this in https://github.com/openhab/openhab-addons/pull/15727/commits/534f593e3b23bc703910aaf04e5464663274eef1 as I could not reproduce this (neither for German nor English locale, and the source code of Calimero does not contain it as well).

File: bundles/org.openhab.binding.knx/doc/dpt.txt [hunk 1]
```diff
@@ -46,8 +46,8 @@ MainType: 3
 3.008: DPT_Control_Blinds values: 0 = up 1 = down
 
 MainType: 4
-4.001: DPT_Char_ASCII
-4.002: DPT_Char_8859_1
+unsupported 4.001: DPT_Char_ASCII
+unsupported 4.002: DPT_Char_8859_1
```
Comment (J-N-K) [inline bundles/org.openhab.binding.knx/doc/dpt.txt:50]: Just a side-note, not related and out of scope: why don't we support that?

File: bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/ValueDecoder.java [hunk 2]
```diff
@@ -134,13 +137,22 @@ public class ValueDecoder {
...
+                    return handleDpt19(translator, data);
                 case "20":
                 case "21":
+                    return handleStringOrDecimal(data, value, preferredType, 8);
                 case "22":
+                    return handleStringOrDecimal(data, value, preferredType, 16);
+                case "16":
                 case "28":
+                case "250": // Map all combined color transitions to String,
+                case "252": // as there is no native support is planned.
```
Comment (J-N-K) [inline bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/ValueDecoder.java:149]:
```suggestion
                case "252": // as no native support is planned.
```

File: bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/itests/Back2BackTest.java [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+        helper("253.600", new byte[] { 0x0f, 0x0e, 0x0d, 0x7 },
+                new StringType("x increase 7 steps y increase 6 steps Y increase 5 steps"));
+        // DPT 254.600 DPT_Relative_Control_RGB
+        // cr(1) cg(1) cb(1)
+        helper("254.600", new byte[] { 0x0f, 0x0e, 0x0d },
+                new StringType("R increase 7 steps G increase 6 steps B increase 5 steps"));
+    }
+
+    @Test
+    @AfterAll
```
Comment (J-N-K) [inline bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/itests/Back2BackTest.java:542]: I don't think `@AfterAll` is the correct annotation here. What do you wanzt to achieve with that?

File: bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/itests/Back2BackTest.java [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+import tuwien.auto.calimero.GroupAddress;
+import tuwien.auto.calimero.KNXException;
+import tuwien.auto.calimero.datapoint.CommandDP;
+import tuwien.auto.calimero.datapoint.Datapoint;
+import tuwien.auto.calimero.dptxlator.TranslatorTypes;
+import tuwien.auto.calimero.process.ProcessCommunicator;
+import tuwien.auto.calimero.process.ProcessCommunicatorImpl;
+
+/**
+ * Integration test to check conversion from raw KNX frame data to OH data types and back.
```
Comment (J-N-K) [inline bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/itests/Back2BackTest.java:63]: This is not an integration test in the sense we usually use it in the openHAB context. We call a test an integration test if it starts the whole framework and executes tests in a running application.

**Review by J-N-K — CHANGES_REQUESTED** [review]
In general looks good an very useful.

File: bundles/org.openhab.binding.knx/doc/dpt.txt (hunk 1 above)
Comment (holgerfriedrich) [reply bundles/org.openhab.binding.knx/doc/dpt.txt:50]: Calimero does not support it...

File: bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/itests/Back2BackTest.java (hunk 4 above)
Comment (holgerfriedrich) [reply bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/itests/Back2BackTest.java:63]: I see that as well. Looking from the perspective of the binding code, it is more than a unit test, as we involve the full Calimero stack. How would you call it? Back2Back? E2E? Though, E2E is typically considered even more as integration test....

File: bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/itests/Back2BackTest.java (hunk 3 above)
Comment (holgerfriedrich) [reply bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/itests/Back2BackTest.java:542]: If I got this correctly, 
> @ AfterAll is used to signal that the annotated method should be executed after all tests in the current test class. 

This method should run after all other tests, as each test adds all tested DPTs to ```dptTested```.

File: bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/itests/Back2BackTest.java (hunk 4 above)
Comment (holgerfriedrich) [reply bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/itests/Back2BackTest.java:63]: @J-N-K Do you have a proposal? Or should we merge as it is?

**Comment by holgerfriedrich** [comment]
@kaikreuzer I would like to see this in the next milestone build if possible. It should be merged as 2 separate commits to keep the records. I think we have only one open point above: can we call this an integration test or shall I change the naming to something different.... WDYT?

**Comment by kaikreuzer** [comment]
I'll leave merging this to @J-N-K as he already did the initial review.

**Review by J-N-K — APPROVED** [review]
I would say that the benefit of having these tests outweigh any potential confusion due to the naming. Since no-one seems to have a better proposal, we should merge as-is.

**Comment by holgerfriedrich** [comment]
thanks you!

**Pull request merged**